### PR TITLE
executor: fix bug about JSON_EXTRACT fails to cast as bool

### PR DESCRIPTION
**Pull request by tsthght**
### What problem does this PR solve? 
fix bug about JSON_EXTRACT fails to cast as bool #12233
@ngaut 

### What is changed and how it works?
```
create database testjson;
use testjson;
create table testjson( id int auto_increment not null primary key, j json )default charset=utf8 engine=innodb;
insert into testjson set j='{"test":1}';
select id from testjson where json_extract(j, '$.test');
+----+
| id |
+----+
|  1 |
+----+
1 row in set (0.001 sec)
```
### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -1390,6 +1390,8 @@ func (d *Datum) ToBool(sc *stmtctx.StatementContext) (int64, error) {
 	case KindBinaryLiteral, KindMysqlBit:
 		val, err1 := d.GetBinaryLiteral().ToInt(sc)
 		isZero, err = val == 0, err1
+	case KindMysqlJSON:
+		isZero, err = d.IsNull(), nil
```
Comment (SunRunAway) [inline types/datum.go]: It should be false if `$.test` is zero

File: executor/executor_test.go [hunk 2]
```diff
@@ -1850,6 +1850,12 @@ func (s *testSuiteP1) TestGeneratedColumnRead(c *C) {
 	result = tk.MustQuery(`SELECT * FROM test_gc_read_m`)
 	result.Check(testkit.Rows(`1 2 4`, `2 3 6`))
 
+	// Test for issue #12233
+	tk.MustExec(`create table testjson( id int auto_increment not null primary key, j json )default charset=utf8 engine=innodb;`)
```
Comment (SunRunAway) [inline executor/executor_test.go]: Could you simplify this table schema?

File: executor/executor_test.go [hunk 3]
```diff
@@ -1850,6 +1850,12 @@ func (s *testSuiteP1) TestGeneratedColumnRead(c *C) {
...
+	tk.MustExec(`insert into testjson set j='{"test":1}';`)
+	result = tk.MustQuery(`select id from testjson where json_extract(j, '$.test');`)
+	result.Check(testkit.Rows(`1`))
```
Comment (SunRunAway) [inline executor/executor_test.go]: I've tried to checkout your pull request, build it and run in a local environment, and it returns an error.

```mysql
mysql> select id from testjson where json_extract(j, '$.test');
ERROR 1105 (HY000): InvalidDataType("can\'t convert Json(1) to bool")
```

Could you do an investigation why it works well in unit test but fails in a local environment?

File: executor/executor_test.go [hunk 4]
```diff
@@ -1850,6 +1850,12 @@ func (s *testSuiteP1) TestGeneratedColumnRead(c *C) {
 	result = tk.MustQuery(`SELECT * FROM test_gc_read_m`)
 	result.Check(testkit.Rows(`1 2 4`, `2 3 6`))
 
+	// Test for issue #12233
```
Comment (SunRunAway) [inline executor/executor_test.go]: Please do not put it in `TestGeneratedColumnRead`

File: executor/executor_test.go (hunk 3 above)
Comment (tsthght) [reply executor/executor_test.go]: In my local environment, it works well. Did you checkout this PR? Or describe your test steps in detail.
image

File: executor/executor_test.go (hunk 4 above)
Comment (tsthght) [reply executor/executor_test.go]: thank you. I had removed this comment.

File: executor/executor_test.go (hunk 2 above)
Comment (tsthght) [reply executor/executor_test.go]: Thank you. I had simplified the table schema.

File: types/datum.go (hunk 1 above)
Comment (tsthght) [reply types/datum.go]: I had add testcase as follows:
a) testcase1:
insert into testjson set j='{"test":0}
select j from testjson where json_extract(j, '$.test')

b) testcase2:
insert into testjson set j='{"test1":0}
select j from testjson where json_extract(j, '$.test')

File: executor/executor_test.go (hunk 3 above)
Comment (SunRunAway) [reply executor/executor_test.go]: It's my fault.

File: executor/executor_test.go (hunk 4 above)
Comment (SunRunAway) [reply executor/executor_test.go]: Sorry, maybe you don't get me.

Your test case is in `TestGeneratedColumnRead`, but it should not be here.

File: types/datum.go (hunk 1 above)
Comment (SunRunAway) [reply types/datum.go]: I've tested this case in Mysql5.7, and the result is different from yours.
```mysql
mysql> drop table testjson;
Query OK, 0 rows affected (0.00 sec)

mysql> create table testjson(j json );
Query OK, 0 rows affected (0.01 sec)

mysql> insert into testjson set j='{"test":0}';
Query OK, 1 row affected (0.00 sec)

mysql> select j from testjson where json_extract(j, '$.test');
Empty set (0.00 sec)
```

Comment (SunRunAway) [reply types/datum.go]: Another case is different from Mysql5.7

In your branch:
```mysql
mysql> create table testjson(j json);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into testjson set j='0';
Query OK, 1 row affected (0.00 sec)

mysql> insert into testjson set j='null';
Query OK, 1 row affected (0.00 sec)

mysql> insert into testjson set j='1';
Query OK, 1 row affected (0.00 sec)

mysql> select j from testjson where j;
+------+
| j    |
+------+
| 0    |
| null |
| 1    |
+------+
3 rows in set (0.00 sec)
```

In MySQL5.7
```mysql
mysql> create table testjson(j json);
Query OK, 0 rows affected (0.02 sec)

mysql> insert into testjson set j='0';
Query OK, 1 row affected (0.01 sec)

mysql> insert into testjson set j='null';
Query OK, 1 row affected (0.00 sec)

mysql> insert into testjson set j='1';
Query OK, 1 row affected (0.00 sec)

mysql> select j from testjson where j;
+------+
| j    |
+------+
| 1    |
+------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+---------------------------------------------------------------+
| Level   | Code | Message                                                       |
+---------+------+---------------------------------------------------------------+
| Warning | 3156 | Invalid JSON value for CAST to INTEGER from column j at row 2 |
+---------+------+---------------------------------------------------------------+
1 row in set (0.00 sec)
```

Comment (tsthght) [reply types/datum.go]: Thank you, I will investigate it.

**Comment by Deardrops** [comment]
@tsthght Friendly ping, any updated for this PR?

**Pull request closed without merging**